Add logic for overwrap of micros()

LaPommeCosmique opened the pull request.

LaPommeCosmique committed:
Add logic for overwrap of micros()

LaPommeCosmique commented on src/Sensor/SensorOptical.cpp:
```diff
@@ -11,6 +11,8 @@
 #define ROLLER_RADIUS 0.08276057 // metres
 #define VELOCITY_FACTOR (GEAR_RATIO * ROLLER_RADIUS) // speed of vehicle [m/s] = velocity factor * angular velocity [rad/s]
 
+#define UINT32_T_MAX 0xffffffffU
```
It seems like there is already a definition for ```UINT32_MAX```, but it expands to ```0xffffffffUL``` which I believe would be a 64 bit variable. As mentioned in issue #62 , this is not desirable

cosparks replied:
```unsigned long``` is actually a 32-bit type.  Like I said before, the size of C types vary depending on the architecture.  If you were compiling for a 64-bit CPU, ```unsigned long``` would be a 64 bits, but for the pico it is still 32 bits.  I think the C type for a 64-bit unsigned int on the pico would be ```unsigned long long```.  Just another example of why in our own code we use the explicitly-defined types in <stdint.h>.

In general, I would trust the standard library implementations of all these things, as a lot of work goes into making sure that they are correct for a given architecture.  We should trust that all the min and max values defined in <stdint.h> are the correct values/types and all that.

When in doubt, you can print out the size of a type:

```
Serial.println("sizeof(unsigned long): " + String(sizeof(unsigned long));

// prints
sizeof(unsigned long): 4
```

cosparks replied:
If you want to go deeper, you can navigate to the file where all these are defined:

```C:/<user>/.platformio/packages/toolchain-rp2040-earlephilhower/arm-none-eabi/include/stdint.h```

These numeric type definitions and additional macros are part of the arm implementation of the C standard library, designed for the architecture that we're using and included in the platformIO arduino-pico dev kit.

LaPommeCosmique replied:
Sounds good, I have changed it so we are using the system defined ```UINT32_MAX``` macro.

LaPommeCosmique replied:
I have added some unit tests, and my tests did not expose any issues with the logic.

I tested the two cases where micros overflowed and the readInterval time had elapsed since the last update time (in which case angular velocity should be calculated), and two cases where the readInterval time had not elapsed since the last update time (in which case angular velocity should not have been calculated).

LaPommeCosmique committed:
Replace UINT32_T_MAX with system defined macro

LaPommeCosmique committed:
Add tests for optical sensor overflow logic

LaPommeCosmique commented on src/Sensor/SensorOptical.cpp:
```diff
@@ -32,11 +32,36 @@ void SensorOptical::begin() {
 void SensorOptical::handle() {
     uint32_t currentTime = micros();
-    if (currentTime >= _lastUpdateTime + _readInterval) {
+    uint32_t nextUpdateTime = _lastUpdateTime + _readInterval;
```
I'm not sure how much computation is required if we calculate ```nextUpdateTime``` and ```overflow``` each time we call ```handle()```. If this uses too much overhead, we could keep them as ```SensorOptical``` properties, since they only change during updates.

cosparks reviewed (commented):
I like this method for predicting overflow.  However, there is another case to consider for the tests, where SensorOptical sets the next read time to <= UINT32_MAX, but handle gets called after micros has overflowed:

```
SECTION("micros overflow -- SensorOptical doesn't expect overflow") {
    setMicros(UINT32_MAX - 10000);
    optical.begin();
    setMicros(0);
    simulateAperturePasses(optical, SensorOptical::NumApertures / 16);
    REQUIRE( optical.getAngularVelocity() == Approx(39.268).margin(0.01) );
}
```

I'm not sure how easy it would be to modify your code to account for this case.  A simpler method (which I think works) would be to just check if ```currentTime < _lastUpdateTime```:

```
    uint32_t deltaT;
    if (currentTime < _lastUpdateTime) {
        // overflow: compute interval with overflow
        deltaT = UINT32_MAX - _lastUpdateTime + currentTime + 1;
    } else {
        // no overflow
        deltaT = currentTime - _lastUpdateTime;
    }
    if (deltaT >= _readInterval) {
        invokeUpdate = true;
    }
```

*Edit: you wouldn't need to branch on ```invokeUpdate``` in this case, as ```deltaT >= _readInterval``` is the condition

LaPommeCosmique committed:
Revise micros wrap check

LaPommeCosmique committed:
Add SensorOptical overflow check

LaPommeCosmique commented:
> 



> I like this method for predicting overflow. However, there is another case to consider for the tests, where SensorOptical sets the next read time to <= UINT32_MAX, but handle gets called after micros has overflowed:
> 
> ```
> SECTION("micros overflow -- SensorOptical doesn't expect overflow") {
>     setMicros(UINT32_MAX - 10000);
>     optical.begin();
>     setMicros(0);
>     simulateAperturePasses(optical, SensorOptical::NumApertures / 16);
>     REQUIRE( optical.getAngularVelocity() == Approx(39.268).margin(0.01) );
> }
> ```
> 
> I'm not sure how easy it would be to modify your code to account for this case. A simpler method (which I think works) would be to just check if `currentTime < _lastUpdateTime`:
> 
> ```
>     uint32_t deltaT;
>     if (currentTime < _lastUpdateTime) {
>         // overflow: compute interval with overflow
>         deltaT = UINT32_MAX - _lastUpdateTime + currentTime + 1;
>     } else {
>         // no overflow
>         deltaT = currentTime - _lastUpdateTime;
>     }
>     if (deltaT >= _readInterval) {
>         invokeUpdate = true;
>     }
> ```
> 
> *Edit: you wouldn't need to branch on `invokeUpdate` in this case, as `deltaT >= _readInterval` is the condition

Thats a good catch, I didn't think of that. I replaced the micros wrap check and added a new unit test and it looks like the current logic passes all tests now.

cosparks commented:
> > 
> 
> > I like this method for predicting overflow. However, there is another case to consider for the tests, where SensorOptical sets the next read time to <= UINT32_MAX, but handle gets called after micros has overflowed:
> > ```
> > SECTION("micros overflow -- SensorOptical doesn't expect overflow") {
> >     setMicros(UINT32_MAX - 10000);
> >     optical.begin();
> >     setMicros(0);
> >     simulateAperturePasses(optical, SensorOptical::NumApertures / 16);
> >     REQUIRE( optical.getAngularVelocity() == Approx(39.268).margin(0.01) );
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'm not sure how easy it would be to modify your code to account for this case. A simpler method (which I think works) would be to just check if `currentTime < _lastUpdateTime`:
> > ```
> >     uint32_t deltaT;
> >     if (currentTime < _lastUpdateTime) {
> >         // overflow: compute interval with overflow
> >         deltaT = UINT32_MAX - _lastUpdateTime + currentTime + 1;
> >     } else {
> >         // no overflow
> >         deltaT = currentTime - _lastUpdateTime;
> >     }
> >     if (deltaT >= _readInterval) {
> >         invokeUpdate = true;
> >     }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > *Edit: you wouldn't need to branch on `invokeUpdate` in this case, as `deltaT >= _readInterval` is the condition
> 
> Thats a good catch, I didn't think of that. I replaced the micros wrap check and added a new unit test and it looks like the current logic passes all tests now.

Cool!  If it passes that test then I think we're good to go.  Nice work
